conduit/pipeline/pipeline.go

```diff
@@ -431,8 +455,228 @@ func addMetrics(block data.BlockData, importTime time.Duration) {
+				p.cancelWithProblem(err)
+				p.logger.Error(err)
+			}
+		}()
+
+		for {
+			p.logger.Tracef("exporter handler waiting for block data. round=%d", lastRound)
+			selectStart := time.Now()
+			select {
+			case <-p.ctx.Done():
+				p.logger.Infof("exporter handler exiting lastRnd=%d", lastRound)
+				return
+			case blkData := <-blkChan:
+				waitTime := time.Since(selectStart)
+				totalSelectWait += waitTime
+				p.logger.Tracef("exporter handler received block data for round %d after wait of %s", lastRound, waitTime)
+				lastRound = blkData.Round()
+
+				if p.pipelineMetadata.NextRound != lastRound {
+					lastError = fmt.Errorf("aborting after out of order block data. %d != %d", p.pipelineMetadata.NextRound, lastRound)
+					return
+				}
+
+				var exportTime time.Duration
+				exportTime, lastError = RetriesNoOutput(exporter.Receive, blkData, p, exporter.Metadata().Name)
+				if lastError != nil {
+					lastError = fmt.Errorf("aborting after failing to export round %d: %w", lastRound, lastError)
+					return
+				}
+				metrics.ExporterTimeSeconds.Observe(exportTime.Seconds())
+				// Ignore round 0 (which is empty).
+				if lastRound > 0 {
+					// Previously we reported time starting after block fetching is complete
+					// through the end of the export operation. Now that each plugin is running
+					// in its own goroutine, report only the time of the export.
+					addMetrics(blkData, exportTime)
+				}
+
+				// Increment Round, update metadata
+				nextRound := lastRound + 1
+				p.pipelineMetadata.NextRound = nextRound
+				_, lastError = RetriesNoOutput(p.pipelineMetadata.encodeToFile, p.cfg.ConduitArgs.ConduitDataDir, p, "pipelineMetadata save")
+				if lastError != nil {
+					lastError = fmt.Errorf("aborting after updating NextRound=%d and failing to save metadata: %w", nextRound, lastError)
+					return
+				}
+				p.logger.Tracef("exporter %s @ round %d saved pipeline metadata", exporter.Metadata().Name, p.pipelineMetadata.NextRound)
+
+				for i, cb := range p.completeCallback {
+					p.logger.Tracef("exporter %s @ round=%d NextRound=%d executing callback %d", exporter.Metadata().Name, lastRound, p.pipelineMetadata.NextRound, i)
+					_, lastError = RetriesNoOutput(cb, blkData, p, fmt.Sprintf("callback %d", i))
+					if lastError != nil {
+						lastError = fmt.Errorf("aborting due to failed callback %d: %w", i, lastError)
+						return
+					}
+				}
+				lastError = nil
+				// WARNING: removing the following will BREAK:
+				// - the E2E test (Search for "Pipeline round:" in subslurp.py)
+				// - the internal tools logstats collector (See func ConduitCollector in logstats.go of internal-tools repo)
+				// Modify with CAUTION!!!!
+				p.logger.Infof("UPDATED Pipeline NextRound=%d. FINISHED Pipeline round r=%d (%d txn) exported in %s", nextRound, lastRound, len(blkData.Payset), time.Since(startStart))
```

Comment:
Here's a [playground](https://go.dev/play/p/_idxxpGBlYL) showing that this new log message should still work with the `logstats.go` regexp.

Reply:
is the time.Since(startStart) correct after the 1st block? `startStart` is the time when the pipeline started and it never gets updated, right?

Reply:
This is a good point. In fact, the [original code](https://github.com/algorand/conduit/blob/master/conduit/pipeline/pipeline.go#L474-L498) was logging the time since the importer was completed, not the time from the beginning of `Start()`. Unfortunately, now that channels intermediate between plugins we can't easily access this value as we did in the original code. If this statistic is important enough, instead of having the channels be of type `chan data.BlockData` they could be of type `chan struct{importFinishTime time.Time, blk data.BlockData}`.  What do you think?

Reply:
or we can add a timestamp data.BlockData?

Reply:
Maybe embed new `type PipelineData struct ...` inside of `BlockData` ?

Reply:
4103de1c50b8ef0c89af9ec3197437b9808b067f

Reply:
I had a change of mind. I don't want to embed `PipelineData` inside of `data.BlockData`. That is mixing together 2 totally different things and even has a risk of "infecting" the indexer database (e.g. suppose the queries naively serialize the entirety of `data.BlockData` into json and just save as is...). So I'll be moving `PipelineData` into the `pipeline` package and creating something like `type block struct` which itself embeds a `data.BlockData`.

Reply:
5ae8a21baaabd327d143c18f68efa79338ef3338